include/nanobind/eigen/sparse.h

```diff
@@ -0,0 +1,124 @@
+            if (!obj.type().is(matrix_type))
+                obj = matrix_type(obj);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        if (object data_o = obj.attr("data"); !scalar_caster.from_python(data_o, flags, cleanup))
+            return false;
+        ScalarTensor& values = scalar_caster.value;
+
+        if (object indices_o = obj.attr("indices"); !storageindex_caster.from_python(indices_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& inner_indices = storageindex_caster.value;
+
+        if (object indptr_o = obj.attr("indptr"); !scalar_caster.from_python(indptr_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& outer_indices = storageindex_caster.value;
+
+        object shape_o = obj.attr("shape"), nnz_o = obj.attr("nnz");
+        Index rows, cols, nnz;
+        try {
+            if (len(shape_o) != 2)
+                return false;
+            rows = cast<Index>(shape_o[0]);
+            cols = cast<Index>(shape_o[1]);
+            nnz = cast<Index>(nnz_o);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        value = EigenMapSparseMatrix<Scalar,
+                                     T::Flags &(Eigen::RowMajor | Eigen::ColMajor),
+                                     StorageIndex>(rows,
+                                                   cols,
+                                                   nnz,
+                                                   outer_indices.data(),
+                                                   inner_indices.data(),
+                                                   values.data());
+
+        return true;
+    }
+
+    static handle from_cpp(T &&v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        if (policy == rv_policy::automatic ||
+            policy == rv_policy::automatic_reference)
+            policy = rv_policy::move;
+
+        return from_cpp((const T &) v, policy, cleanup);
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        // FIXME: Looks like UB. Is there anything that can be done about this?
+        T & src = const_cast<T &>(v);
+        src.makeCompressed();
+
+        object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+
+        // FIXME: These casts are ugly! Is there another way to write this?
+        const size_t data_shape[] = { static_cast<size_t>(v.nonZeros()) };
+        const size_t outerIndices_shape[] = { static_cast<size_t>((row_major ? v.rows() : v.cols()) + 1) };
+        ScalarTensor data(src.valuePtr(), 1, data_shape);
+        StorageIndexTensor outer_indices(src.outerIndexPtr(), 1, outerIndices_shape);
+        StorageIndexTensor inner_indices(src.innerIndexPtr(), 1, data_shape);
+
+        return matrix_type(make_tuple(
```

Comment:
This function call can fail with an exception, and that is not allowed by a caster. Once more, I think there is a whole bunch of code that must be wrapped:

```cpp
try {
    // code that may throw
} catch (const python_error &e) {
    e.restore();
    return handle();
}
```